File: core/src/net/artux/pda/map/DataRepository.java
package net.artux.pda.map;

import net.artux.pda.map.utils.PlatformInterface;
import net.artux.pda.model.QuestUtil;
import net.artux.pda.model.map.GameMap;
import net.artux.pda.model.quest.StoryModel;
import net.artux.pda.model.quest.story.StoryDataModel;
import net.artux.pda.model.user.UserModel;

import org.apache.commons.lang3.SerializationUtils;

import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;


public class DataRepository {

    private PropertyChangeSupport propertyChangeSupport;
    private StoryDataModel oldStoryDataModel;
    private StoryDataModel storyDataModel;

    private GameMap gameMap;
    private StoryModel storyModel;
    private UserModel userModel;

    private Properties properties;
    private final PlatformInterface platformInterface;

    public DataRepository(PlatformInterface platformInterface) {
        this.platformInterface = platformInterface;
        this.properties = new Properties();
    }

    public void addPropertyChangeListener(PropertyChangeListener listener) {
        propertyChangeSupport.addPropertyChangeListener(listener);
        setStoryDataModel(storyDataModel);
    }

    public void removePropertyChangeListener(PropertyChangeListener listener) {
        propertyChangeSupport.removePropertyChangeListener(listener);
    }

    public void setProperties(Properties properties) {
        this.properties = properties;
    }

    public Properties getProperties() {
        return properties;
    }

    public StoryDataModel getStoryDataModel() {
        return storyDataModel;
    }

    public void setStoryDataModel(StoryDataModel storyDataModel) {
        if (propertyChangeSupport == null) {
            propertyChangeSupport = new PropertyChangeSupport(storyDataModel);
        }
        if (oldStoryDataModel == null) {
            oldStoryDataModel = storyDataModel;
        } else
            oldStoryDataModel = this.storyDataModel;
        this.storyDataModel = SerializationUtils.clone(storyDataModel);
        propertyChangeSupport.firePropertyChange("storyData", oldStoryDataModel, storyDataModel);
    }

    public GameMap getGameMap() {
        return gameMap;
    }

    public void setGameMap(GameMap gameMap) {
        this.gameMap = gameMap;
    }

    public void setStoryModel(StoryModel storyModel) {
        this.storyModel = storyModel;
    }

    public StoryModel getStoryModel() {
        return storyModel;
    }

    public UserModel getUserModel() {
        return userModel;
    }

    public void setUserModel(UserModel userModel) {
        this.userModel = userModel;
    }

    public PlatformInterface getPlatformInterface() {
        return platformInterface;
    }

    public void applyActions(Map<String, List<String>> actions) {
        HashMap<String, List<String>> summaryMap
                = new HashMap<>(QuestUtil.difference(oldStoryDataModel, storyDataModel));
        if (actions != null && actions.size() > 0) {
            summaryMap.putAll(actions);
            platformInterface.applyActions(summaryMap);
        }
    }
}

File: core/src/net/artux/pda/map/DataRepository.kt
package net.artux.pda.map

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import net.artux.pda.map.utils.PlatformInterface
import net.artux.pda.model.QuestUtil
import net.artux.pda.model.items.ItemsContainerModel
import net.artux.pda.model.map.GameMap
import net.artux.pda.model.quest.StoryModel
import net.artux.pda.model.quest.story.StoryDataModel
import java.util.*


class DataRepository(
    val platformInterface: PlatformInterface,
    private val storyDataModel: StoryDataModel,
    var gameMap: GameMap,
    val items: ItemsContainerModel,
    val storyModel: StoryModel,
    val properties: Properties
) {


    data class Builder(
        var platformInterface: PlatformInterface? = null,
        var storyDataModel: StoryDataModel? = null,
        var gameMap: GameMap? = null,
        var items: ItemsContainerModel? = null,
        var storyModel: StoryModel? = null,
        var properties: Properties? = null
    ) {

        fun platformInterface(platformInterface: PlatformInterface) =
            apply { this.platformInterface = platformInterface }

        fun storyDataModel(storyDataModel: StoryDataModel) =
            apply { this.storyDataModel = storyDataModel }

        fun gameMap(gameMap: GameMap) = apply { this.gameMap = gameMap }
        fun items(items: ItemsContainerModel) = apply { this.items = items }
        fun storyModel(storyModel: StoryModel) = apply { this.storyModel = storyModel }
        fun properties(properties: Properties) = apply { this.properties = properties }
        fun build() = DataRepository(
            platformInterface!!,
            storyDataModel!!,
            gameMap!!,
            items!!,
            storyModel!!,
            properties!!
        )
    }

    var previousStoryDataModel = storyDataModel
    var currentStoryDataModel = storyDataModel
    private val dataModelFlow: MutableSharedFlow<StoryDataModel> = MutableSharedFlow(
        replay = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )
    val storyDataModelFlow: SharedFlow<StoryDataModel> = dataModelFlow

    init {
        setStoryDataModel(storyDataModel)
    }

    fun setStoryDataModel(storyDataModel: StoryDataModel) {
        previousStoryDataModel = currentStoryDataModel
        currentStoryDataModel = storyDataModel
        dataModelFlow.tryEmit(storyDataModel)
    }

    fun sendData(map: Map<String, String>) {
        applyActions(emptyMap())
        platformInterface.send(map)
    }

    fun update() {
        setStoryDataModel(currentStoryDataModel)
    }

    fun applyActions(actions: Map<String, List<String>>?) {
        val summaryMap = HashMap(QuestUtil.difference(previousStoryDataModel, storyDataModel))
        if (actions != null && actions.isNotEmpty()) {
            summaryMap.putAll(actions)
            platformInterface.applyActions(summaryMap)
        }
    }
}